Common/NearestNeighbour/src/main/java/org/tribuo/common/nearest/KNNModel.java

```diff
@@ -107,16 +108,21 @@ public enum Backend {
 
     private final EnsembleCombiner<T> combiner;
 
+    // This is not final to support deserialization of older models. It will be final in a future version which doesn't
+    // maintain serialization compatibility with 4.X.
+    private NeighboursQueryFactory neighboursQueryFactory;
```

Comment:
A transient field for the `NeighboursQuery` object has been added, and is safely constructed on demand.

Reply:
Let's just construct the query in the constructor and `readObject` rather than have the locking in the prediction methods. If you've constructed a `KNNModel` then it's reasonable to assume that users want to use it, and that avoids the synchronisation to construct it lazily.